From fe75c6184acd09e27e22fa4c4f0ba132c39bf5f1 Mon Sep 17 00:00:00 2001
From: Christine Gharzuzi <chrisg@marvell.com>
Date: Wed, 15 Feb 2017 17:47:14 +0200
Subject: [PATCH 0780/1345] dts: a70x0: add device tree configuration for
 board setup C

commit  e40122b773ed1435f3cc23b63160b7c4eb1b6d21 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

A7040 DB board is a modular board,
which supports multiple possible configurations.
This patch introduce predefined board topology for setup C.

- Add device tree for C setup
	- AP_UART, AP_SD
	- CP0_RGMII1, CP0_SDIO, CP0_I2C,
	  CP0_SATA1, CP0_SPI

Change-Id: I8b73c51bf30f5cda5c5f33557a6b976d950467ba
Signed-off-by: Christine Gharzuzi <chrisg@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36822
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../bindings/arm/marvell/armada-7k-8k.txt          |    3 +
 .../mvebu/a7k-a8k/armada7040-db-setup.txt          |  126 ++++++-------
 arch/arm64/boot/dts/marvell/Makefile               |    1 +
 arch/arm64/boot/dts/marvell/armada-7040-db-C.dts   |  186 ++++++++++++++++++++
 arch/arm64/boot/dts/marvell/armada-7040.dtsi       |    6 +-
 arch/arm64/boot/dts/marvell/armada-cp110.dtsi      |   12 ++
 6 files changed, 273 insertions(+), 61 deletions(-)
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db-C.dts

diff --git a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
index a35c6e1..805cc00 100644
--- a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
+++ b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
@@ -26,6 +26,9 @@ Boards:
    - "marvell,armada7040-db-B", "marvell,armada7040-db"
       when the board being used is the Armada 7040 Development board
 
+   - "marvell,armada7040-db-C", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board
+
    - "marvell,armada7040-db-nand", "marvell,armada7040-db"
       when the board being used is the Armada 7040 Development board
 
diff --git a/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
index 8e70fc7..bba3b5f 100644
--- a/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
+++ b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
@@ -18,36 +18,24 @@ To switch board configuration:
 
 Board Setup
 ------------
-
 For A7040 A revision SoC - Please use armada-7040-db-A configuration,
 
-DPR Setting
-------------
-|					| 	1: A				 |
-|---------------------------------------|----------------------------------------|
-| Mpp Jumper Module (SLM1548)		| Default release (ALL DPRs 1-2 position)|
-| Serdess Jumper Module (SLM1547)	| DPR152,153,154,155 - 1-2 position	 |
-|					| DPR148,149,150,151 - 2-3 position	 |
-| CP0 Serdess module (SLM1521)		| SW1 - (1,3,6,7)= ON,  (2,4,5,8)=OFF	 |
-| 					| DPR2,3,4,5 - 1-2 position		 |
-----------------------------------------|----------------------------------------|
-
-|		| 	1: A		|	2:B		|
-|---------------|-----------------------|-----------------------|
-|Device tree	|armada-7040-db-A.dtb	|armada-7040-db-B.dtb	|
-|CP0-SW1:1	| ON			| OFF			|
-|CP0-SW1:2	| OFF			| ON			|
-|CP0-SW1:3	| ON			| ON			|
-|CP0-SW1:4	| OFF			| ON			|
-|CP0-SW1:5	| OFF			| ON			|
-|CP0-SW1:6	| ON			| ON			|
-|CP0-SW1:7	| ON			| ON			|
-|CP0-SW1:8	| OFF			| OFF			|
------------------------------------------------------------------
+|		| 	1: A		|	2:B		|	3:C		|
+|---------------|-----------------------|-----------------------|-----------------------|
+|Device tree	|armada-7040-db-A.dtb	|armada-7040-db-B.dtb	|armada-7040-db-C.dtb	|
+|CP0-SW1:1	| ON			| OFF			| ON			|
+|CP0-SW1:2	| OFF			| ON			| OFF			|
+|CP0-SW1:3	| ON			| ON			| ON			|
+|CP0-SW1:4	| OFF			| ON			| OFF			|
+|CP0-SW1:5	| OFF			| ON			| OFF			|
+|CP0-SW1:6	| ON			| ON			| ON			|
+|CP0-SW1:7	| ON			| ON			| ON			|
+|CP0-SW1:8	| OFF			| OFF			| OFF			|
+-----------------------------------------------------------------------------------------
 
 DPR Setting
 ------------
-|					| 	1: A				 |	2: B				  |
+|					| 	1: A				 |	2: B:                             |
 |---------------------------------------|----------------------------------------|----------------------------------------|
 | Mpp Jumper Module (SLM1548)		| Default release (ALL DPRs 1-2 position)| Default release (ALL DPRs 1-2 position)|
 | Serdess Jumper Module (SLM1547)	| DPR152,153,154,155 - 1-2 position	 | DPR152,153,154,155 - 1-2 position      |
@@ -56,20 +44,30 @@ DPR Setting
 | 					| DPR2,3,4,5 - 1-2 position		 | DPR2,3,4,5 - 2-3 position              |
 ----------------------------------------|----------------------------------------|----------------------------------------|
 
+|					| 	3: C				 |
+|---------------------------------------|----------------------------------------|
+| Mpp Jumper Module (SLM1548)		| DPR133 - 143  2-3 position 		 |
+| Serdess Jumper Module (SLM1547)	| DPR152,153,154,155 - 1-2 position	 |
+|					| DPR148,149,150,151 - 1-2 position	 |
+| CP0 Serdess module (SLM1521)		| SW1 - (1,3,6,7)= ON,  (2,4,5,8)=OFF	 |
+| 					| DPR2,3,4,5 - 1-2 position		 |
+| Install TDM Module (SLM1448)		| on DB-7040 CON46			 |
+----------------------------------------|----------------------------------------|
+
 The tables below summarizes the interface configuration of each setup
 
 SerDes Configuration
 ------------------------
 
-| CP0 Lane 	| 	1: A		|	2: B		|
-|---------------|-----------------------|-----------------------|
-| 0		| SGMII2		|	PICE0 (x4)	|
-| 1		| USB3_HOST0		|	PCIE0 (x4)	|
-| 2		| KR (10G)		|	PCIE0 (x4)	|
-| 3		| SATA1			|	PCIE0 (x4)	|
-| 4		| USB3_HOST1		|	USB3_HOST1	|
-| 5		| PCIE2 (x1)		|	PCIE2 (x1)	|
------------------------------------------------------------------
+| CP0 Lane 	| 	1: A		|	2: B		|       3: C            |
+|---------------|-----------------------|-----------------------|-----------------------|
+| 0		| SGMII2		|	PICE0 (x4)	|       PCIE0 (x1)      |
+| 1		| USB3_HOST0		|	PCIE0 (x4)	|       USB3_HOST0      |
+| 2		| SFI (10G)		|	PCIE0 (x4)	|       SFI (10G)      	|
+| 3		| SATA1			|	PCIE0 (x4)	|       SATA1		|
+| 4		| USB3_HOST1		|	USB3_HOST1	|       USB3_HOST1      |
+| 5		| PCIE2 (x1)		|	PCIE2 (x1)	|       PCIE2 (x1)      |
+-----------------------------------------------------------------------------------------
 
 - USB2_H0 (UTMI only)
 - USB2_H1 (UTMI & USB3 SerDes)
@@ -77,30 +75,31 @@ SerDes Configuration
 
 Multi-purpose pin configurations
 --------------------------------
-Setup  #	|	1	|	2	|
-Name   #	|  	A 	|	B	|
--------------------------------------------------
-AP806:				|		|
--------------------------------------------------
-	AP-SDIO | [0-5]		| [0-5]		|
-	AP-UART0| [11,19]	| [11,19]	|
--------------------------------------------------
-CP:				|		|
--------------------------------------------------
-	CP-SPI0	| N/C		| N/C		|
-	CP-SPI1	| [13-16]	| [13-16]	|
-	NAND	| N/C		| N/C		|
-	RGMII0	| N/C		| [44-55]	|
-	RGMII1	| [0-11]	| [0-11]	|
-	CP-UART0| [29-30]	| [29-30]	|
-	CP-UART1| N/C		| N/C		|
-	SD	| [56-61]	| [56-61]	|
-	GPIO	| [62]		| [62]		|
-	TDM	| N/C		| N/C		|
-	TWSI	| [37-38]	| [37-38]	|
-	SATA0	| N/C		| N/C		|
-	SATA1	| [28]		| [28]		|
--------------------------------------------------
+
+Setup  #	|	1	|	2	|       3       |
+Name   #	|  	A 	|	B	|       C       |
+-----------------------------------------------------------------
+AP806:				|		|		|
+-----------------------------------------------------------------
+	AP-SDIO | [0-5]		| [0-5]		| [0-5]         |
+	AP-UART0| [11,19]	| [11,19]	| [11,19]       |
+-----------------------------------------------------------------
+CP:				|		|		|
+-----------------------------------------------------------------
+	CP-SPI0	| N/C		| N/C		| [6-11]	|
+	CP-SPI1	| [13-16]	| [13-16]	| [13-16]	|
+	NAND	| N/C		| N/C		| N/C		|
+	RGMII0	| N/C		| [44-55]	| [0-11]	|
+	RGMII1	| [0-11]	| [0-11]	| [44-55]	|
+	CP-UART0| [29-30]	| [29-30]	| [29-30	|
+	CP-UART1| N/C		| N/C		| N/C		|
+	SD	| [56-61]	| [56-61]	| [56-61]	|
+	GPIO	| [62]		| [62]		| [62]		|
+	TDM	| N/C		| N/C		| [0-5]		|
+	TWSI	| [37-38]	| [37-38]	| [37-38]	|
+	SATA0	| N/C		| N/C		| N/C		|
+	SATA1	| [28]		| [28]		| [28]		|
+-----------------------------------------------------------------
 
 Network configuration
 ---------------------
@@ -115,6 +114,13 @@ Network configuration
 2: B:
  Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
  ---------------|-------|---------------|---------------|-------------------------------|
- eth0           | CP0   | eth1          | 2             | RGMII0                        |
- eth1           | CP0   | eth2          | 3             | RGMII1                        |
+ eth0           | CP0   | eth0          | 2             | RGMII0                        |
+ eth1           | CP0   | eth1          | 3             | RGMII1                        |
+ ----------------------------------------------------------------------------------------
+
+ 3: C:
+ Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0           | CP0   | eth0          | 0             | SFI (10G)                     |
+ eth1           | CP0   | eth1          | 3             | RGMII1                        |
  ----------------------------------------------------------------------------------------
diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index 21ad31d..38f9ec1 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -12,6 +12,7 @@ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-A.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-B.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-C.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-nand.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts b/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts
new file mode 100644
index 0000000..c33662b
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2016 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for Marvell Armada 7040 development board
+ * This board file supports the C configuration of the board
+ */
+
+#include "armada-7040-db.dtsi"
+
+/ {
+	model = "Marvell Armada-7040 development board C setup";
+	compatible = "marvell,armada7040-db-C", "marvell,armada7040-db", "marvell,armada7040",
+			"marvell,armada-ap806-quad", "marvell,armada-ap806";
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	chosen { };
+
+	ap806 {
+		config-space {
+			sdhci@6e0000 {
+				status = "okay";
+			};
+			serial@512000 {
+				status = "okay";
+			};
+			i2c@511000 {
+				status = "disabled";
+				clock-frequency = <100000>;
+			};
+		};
+	};
+
+	cpn-110-master {
+		config-space {
+			serial@702000 {
+				status = "disabled";
+			};
+			serial@702100 {
+				status = "disabled";
+			};
+			serial@702200 {
+				status = "disabled";
+			};
+			serial@702300 {
+				status = "disabled";
+			};
+			sata: sata@540000 {
+				status = "okay";
+			};
+			usb3h0: usb3@500000 {
+				status = "okay";
+			};
+			usb3h1: usb3@510000 {
+				status = "okay";
+			};
+			u3d@520000 {
+				status = "disabled";
+			};
+			udc@524100 {
+				status = "disabled";
+			};
+			sdhci@780000 {
+				status = "okay";
+			};
+			spi@700680 {
+				status = "okay";
+			};
+			spi@700600 {
+				status = "okay";
+				slice@0{
+					#address-cells = <1>;
+					#size-cells = <0>;
+					compatible = "mv_slic";
+					reg = <0>;
+					spi-max-frequency = <3000000>;
+					spi-1byte-cs;
+					spi-cpol;
+					spi-cpha;
+				};
+			};
+			tdm@7a0000 {
+				pinctrl-0 = <&cpm_tdm_pins>;
+				pinctrl-names = "default";
+				pclk-freq-mhz = <8>;
+				status = "okay";
+			};
+			i2c1: i2c@701000 {
+				status = "okay";
+				clock-frequency = <100000>;
+			};
+			i2c2: i2c@701100 {
+				status = "disabled";
+			};
+
+			mdio@12a200 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				phy3: ethernet-phy@3 {
+					reg = <1>;
+				};
+			};
+			gop {
+				emac0: mac0 {
+					phy-mode = "sfi"; /*lane-2*/
+					phys = <&comphy0 2 COMPHY_SFI>;
+					phy-names = "comphy";
+				};
+				emac2: mac2 {
+					phy-mode = "none";
+				};
+				emac3: mac3 {
+					phy-mode = "rgmii"; /* rgmii-1 */
+					phy = <&phy3>;
+				};
+			};
+			ppv22@000000 {
+				eth0: eth0@010000 {
+					status = "okay";
+				};
+				eth1: eth1@020000 {
+					status = "disabled";
+				};
+				eth2: eth2@030000 {
+					status = "okay";
+				};
+			};
+			eip197: eip197@800000 {
+				status = "okay";
+			};
+			comphy {
+				status = "okay";
+			};
+		};
+
+		pcie@0x640000 {
+			status = "okay";
+		};
+		pcie@0x620000{
+			status = "disabled";
+		};
+		pcie@0x600000{
+			status="disabled";
+		};
+	};
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-7040.dtsi b/arch/arm64/boot/dts/marvell/armada-7040.dtsi
index d0f4a0e..d95a3bd 100644
--- a/arch/arm64/boot/dts/marvell/armada-7040.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-7040.dtsi
@@ -82,7 +82,11 @@
 		config-space@f2000000 {
 			pinctrl@440000 {
 			compatible = "marvell,a70x0-pinctrl";
-
+				cpm_tdm_pins: tdm-pins {
+					marvell,pins = "mpp0", "mpp1", "mpp2",
+					"mpp3", "mpp4", "mpp5";
+					marvell,function = "tdm";
+				};
 				i2c0_pins: i2c-pins-0 {
 					marvell,pins = "mpp37", "mpp38";
 					marvell,function = "i2c0";
diff --git a/arch/arm64/boot/dts/marvell/armada-cp110.dtsi b/arch/arm64/boot/dts/marvell/armada-cp110.dtsi
index 5137c50..e7bb831 100644
--- a/arch/arm64/boot/dts/marvell/armada-cp110.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-cp110.dtsi
@@ -94,6 +94,18 @@ gpio@440140 {
 		   <ICU_GRP_NSR 79 IRQ_TYPE_LEVEL_HIGH>;
 };
 
+tdm@7a0000 {
+	compatible = "marvell,armada-a8k-tdm";
+	reg = <0x7a0000 0x20000>;
+	reg-names = "tdm_regs";
+	interrupts = <ICU_GRP_NSR 111 IRQ_TYPE_LEVEL_HIGH>,
+		     <ICU_GRP_NSR 112 IRQ_TYPE_LEVEL_HIGH>,
+		     <ICU_GRP_NSR 113 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&cpm_syscon0 1 1>;
+	clock-names = "gateclk";
+	status = "disabled";
+};
+
 serial@702000 {
 	compatible = "snps,dw-apb-uart";
 	reg = <0x702000 0x100>;
-- 
1.7.9.5

